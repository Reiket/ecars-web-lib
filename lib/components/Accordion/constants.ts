export const ACCORDION_TEST_ID = 'accordionTestId';
