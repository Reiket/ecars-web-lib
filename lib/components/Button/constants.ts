import type {ButtonHTMLAttributes} from 'react';

export const BUTTON_SIZE = Object.freeze({
  BIG: 'big',
  SMALL: 'small',
});

export const BUTTON_COLOR = Object.freeze({
  GREEN: 'green',
  WHITE: 'white',
  GRAY: 'gray',
});


export type ButtonAttributesType = ButtonHTMLAttributes<HTMLButtonElement>['type']
export type ButtonSizeType = (typeof BUTTON_SIZE)[keyof typeof BUTTON_SIZE];
export type ButtonColorType = (typeof BUTTON_COLOR)[keyof typeof BUTTON_COLOR];

export const BUTTON_TEST_ID = 'buttonTestId';