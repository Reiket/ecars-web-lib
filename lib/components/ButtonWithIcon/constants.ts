export const LEFT_ICON_BUTTON_TEST_ID = 'leftIconButtonTestId';
export const RIGHT_ICON_BUTTON_TEST_ID = 'rightIconButtonTestId';
