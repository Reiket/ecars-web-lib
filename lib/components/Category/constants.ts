export const CATEGORIES_LIST = Object.freeze({
  GUIDES: 'guides',
  REVIEWS: 'reviews',
  NEWS: 'news',
});

export type CategoriesListType = (typeof CATEGORIES_LIST)[keyof typeof CATEGORIES_LIST];

export const CATEGORY_TEST_ID = 'categoryTestId';
