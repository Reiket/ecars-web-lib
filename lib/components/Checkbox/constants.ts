import type {ElementProps} from '@/services/types';

export interface CheckboxProps extends ElementProps {
  name?: string;
  id?: string;
  disabled?: boolean;
  onChange?: (checked: boolean) => void;
  checked?: boolean;
  hasError?: boolean;
}

export const CHECKBOX_TEST_ID = 'checkboxTestId';
export const CHECKBOX_TEST_INPUT_ID = 'checkboxTestInputId';
