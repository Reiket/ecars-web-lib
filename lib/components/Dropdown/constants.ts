import type {ReactNode} from 'react';

export interface DropdownOption {
  value: string;
  label?: ReactNode;
}

export interface DropdownProps {
  onSelect: (value: string) => void;
  category: string;
  options: DropdownOption[];
}

export const DROPDOWN_TEST_ID = 'dropdownTestId';

export const DROPDOWN_THEME = Object.freeze({
  DARK: 'dark',
  LIGHT: 'light',
});

export type DropdownThemeType = (typeof DROPDOWN_THEME)[keyof typeof DROPDOWN_THEME];
