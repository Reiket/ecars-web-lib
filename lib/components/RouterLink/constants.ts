export const ROUTER_LINK_COLOR = Object.freeze({
  GREEN: 'green',
  GRAY: 'gray',
  DARK: 'dark',
  LIGHTGRAY: 'lightgray',
});

export const ROUTER_LINK_SIZE = Object.freeze({
  SEMIBOLD: 'semibold',
  BOLD: 'bold',
});

export const ROUTER_LINK_TEST_ID = 'routerLinkTestId';

export type RouterLinkColorType = (typeof ROUTER_LINK_COLOR)[keyof typeof ROUTER_LINK_COLOR];
export type RouterLinkSizeType = (typeof ROUTER_LINK_SIZE)[keyof typeof ROUTER_LINK_SIZE];
