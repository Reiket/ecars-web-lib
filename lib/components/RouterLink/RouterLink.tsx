import type {FC, ReactNode} from 'react';
import {Link} from 'react-router-dom';
import type {RouterLinkColorType} from '@/components/RouterLink/constants';
import {ROUTER_LINK_TEST_ID} from '@/components/RouterLink/constants';
import {cn} from '@/services/helpers';
import type {ElementProps} from '@/services/types';

export interface RouterLinkProps extends ElementProps {
  color: RouterLinkColorType;
  children: ReactNode;
  to?: string;
  href?: string;
  withIcon?: boolean;
}

export const RouterLink: FC<RouterLinkProps> = ({color, block, href, to, children, withIcon = false}) => {
  const classNames = cn(block, `link`, `link--${color}`, {
    'link--icon': withIcon,
  });
  if (to) {
    return (
      <Link
        data-testid={ROUTER_LINK_TEST_ID}
        to={to}
        className={classNames}
      >
        {children}
      </Link>
    );
  }

  if (href) {
    return (
      <a
        href={href}
        data-testid={ROUTER_LINK_TEST_ID}
        className={classNames}
        target="_blank"
        rel="noopener noreferrer"
      >
        {children}
      </a>
    );
  }
  return null;
};
