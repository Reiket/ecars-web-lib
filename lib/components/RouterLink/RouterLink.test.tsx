import {cleanup, render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import {ROUTER_LINK_COLOR, ROUTER_LINK_TEST_ID} from '@/components/RouterLink/constants';
import {RouterLink} from '@/components/RouterLink/RouterLink';
import {MemoryRouter} from 'react-router';

describe('Router Link component', () => {
  afterEach(cleanup);

  const colors = Object.values(ROUTER_LINK_COLOR);
  const props = {
    color: ROUTER_LINK_COLOR.GRAY,
  };
  colors.forEach((color) => {
    const testName = `renders links with ${color} color`;
    test(testName, () => {
      const {container} = render(
        <RouterLink
          href="/"
          color={color}
        >
          Link
        </RouterLink>,
      );
      const routerLinkElement = screen.getByTestId(ROUTER_LINK_TEST_ID);

      expect(routerLinkElement).toBeInTheDocument();
      expect(routerLinkElement).toHaveClass(`link--${color}`);
      expect(container).toMatchSnapshot();
    });
  });

  test('applies "link-icon" class when withIcon is true', () => {
    const {container} = render(
      <RouterLink
        href="/"
        withIcon
        {...props}
      >
        Link with Icon
      </RouterLink>,
    );
    const routerLinkElement = screen.getByTestId(ROUTER_LINK_TEST_ID);

    expect(routerLinkElement).toHaveClass('link--icon');
    expect(container).toMatchSnapshot();
  });
  test('renders an external link when "href" is provided', () => {
    const {container} = render(
      <RouterLink
        href="https://example.com"
        {...props}
      >
        External Link
      </RouterLink>,
    );
    const routerLinkElement = screen.getByTestId(ROUTER_LINK_TEST_ID);

    expect(routerLinkElement).toBeInTheDocument();
    expect(routerLinkElement).toHaveAttribute('href', 'https://example.com');
    expect(routerLinkElement).toHaveAttribute('target', '_blank');
    expect(routerLinkElement).toHaveAttribute('rel', 'noopener noreferrer');
    expect(container).toMatchSnapshot();
  });
  test('renders an internal link when "to" is provided', () => {
    const {container} = render(
      <MemoryRouter>
        <RouterLink
          to="/internal"
          {...props}
        >
          Internal Link
        </RouterLink>
        ,
      </MemoryRouter>,
    );
    const routerLinkElement = screen.getByTestId(ROUTER_LINK_TEST_ID);
    expect(routerLinkElement).toBeInTheDocument();
    expect(routerLinkElement).toHaveAttribute('href', '/internal');
    expect(container).toMatchSnapshot();
  });
});
