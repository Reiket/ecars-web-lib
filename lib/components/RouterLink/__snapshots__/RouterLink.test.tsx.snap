// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router Link component applies "link-icon" class when withIcon is true 1`] = `
<div>
  <a
    class="link link--gray link--icon"
    data-testid="routerLinkTestId"
    href="/"
    rel="noopener noreferrer"
    target="_blank"
  >
    Link with Icon
  </a>
</div>
`;

exports[`Router Link component renders an external link when "href" is provided 1`] = `
<div>
  <a
    class="link link--gray"
    data-testid="routerLinkTestId"
    href="https://example.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    External Link
  </a>
</div>
`;

exports[`Router Link component renders an internal link when "to" is provided 1`] = `
<div>
  <a
    class="link link--gray"
    data-discover="true"
    data-testid="routerLinkTestId"
    href="/internal"
  >
    Internal Link
  </a>
  ,
</div>
`;

exports[`Router Link component renders links with colors dark 1`] = `
<div>
  <a
    class="link link--dark"
    data-testid="routerLinkTestId"
    href="/"
    rel="noopener noreferrer"
    target="_blank"
  >
    Link
  </a>
</div>
`;

exports[`Router Link component renders links with colors gray 1`] = `
<div>
  <a
    class="link link--gray"
    data-testid="routerLinkTestId"
    href="/"
    rel="noopener noreferrer"
    target="_blank"
  >
    Link
  </a>
</div>
`;

exports[`Router Link component renders links with colors green 1`] = `
<div>
  <a
    class="link link--green"
    data-testid="routerLinkTestId"
    href="/"
    rel="noopener noreferrer"
    target="_blank"
  >
    Link
  </a>
</div>
`;

exports[`Router Link component renders links with colors lightgray 1`] = `
<div>
  <a
    class="link link--lightgray"
    data-testid="routerLinkTestId"
    href="/"
    rel="noopener noreferrer"
    target="_blank"
  >
    Link
  </a>
</div>
`;
